service/src/main/java/org/cbioportal/service/impl/GenePanelServiceImpl.java

```diff
@@ -86,170 +90,139 @@ public GenePanel getGenePanel(String genePanelId) throws GenePanelNotFoundExcept
-                if (molecularProfileSampleIdentifier.getMolecularProfileId().endsWith("_fusion")) {
-                    hasFusions.set(true);
-                    MolecularProfileCaseIdentifier profileCaseIdentifier = new MolecularProfileCaseIdentifier();
-                    profileCaseIdentifier.setMolecularProfileId(molecularProfileSampleIdentifier.getMolecularProfileId().replace("_fusion", "_mutations"));
-                    profileCaseIdentifier.setCaseId(molecularProfileSampleIdentifier.getCaseId());
-                    return profileCaseIdentifier;
-                }
-                return molecularProfileSampleIdentifier;
-            }).collect(Collectors.toSet());
+            .map(MolecularProfileCaseIdentifier::getMolecularProfileId)
+            .collect(toSet());
 
-        List<MolecularProfile> molecularProfiles = molecularProfileRepository.getMolecularProfiles(new ArrayList<>(molecularProfileIds), "SUMMARY");
+        Map<String, Boolean> queriedMolecularProfileCaseIdentifierMap =
+            molecularProfileCaseIdentifiers
+                .stream()
+                .collect(Collectors
+                    .toMap(
+                        datum -> datum.getMolecularProfileId() + datum.getCaseId(),
+                        datum -> true));
 
-        List<String> studyIds = molecularProfiles
+        return fetchGenePanelDataByMolecularProfileIds(molecularProfileIds)
             .stream()
-            .map(MolecularProfile::getCancerStudyIdentifier)
-            .distinct()
-            .collect(Collectors.toList());
+            .filter(datum -> queriedMolecularProfileCaseIdentifierMap.containsKey(keyGenerator.apply(datum)))
+            .collect(toList());
 
-        int totalSampleCount = studyRepository.fetchStudies(studyIds, "SUMMARY")
-            .stream()
-            .mapToInt(CancerStudy::getAllSampleCount)
-            .sum();
-
-        // fetch all data for molecular profiles if the sampleIdentifiers are more than 30% of total sample in queried studies
-        if (updatedMolecularProfileCaseIdentifiers.size() > totalSampleCount * 0.3) {
-            return new Pair<>(true, fetchGenePanelDataByMolecularProfileIds(molecularProfileIds));
-        } else {
-            return new Pair<>(hasFusions.get(), queryFunction.apply(new ArrayList<>(updatedMolecularProfileCaseIdentifiers)));
+    }
+
+    private List<GenePanelData> annotateDataFromSequencedSampleLists(List<GenePanelData> genePanelData,
+                                                                     MolecularProfile molecularProfile) {
+
+        if (molecularProfile.getMolecularAlterationType().equals(MolecularProfile.MolecularAlterationType.MUTATION_EXTENDED) ||
+            molecularProfile.getMolecularAlterationType().equals(MolecularProfile.MolecularAlterationType.FUSION) ||
+            (molecularProfile.getMolecularAlterationType().equals(MolecularProfile.MolecularAlterationType.STRUCTURAL_VARIANT) &&
```

Comment:
You might as well flip these equals to make this if a bit more resilient to NPEs